plugins/catalog/src/api/CatalogClient.ts

```diff
@@ -35,57 +36,63 @@ export class CatalogClient implements CatalogApi {
     this.apiOrigin = apiOrigin;
     this.basePath = basePath;
   }
-  async getLocationById(id: String): Promise<Location | undefined> {
-    const response = await fetch(
-      `${this.apiOrigin}${this.basePath}/locations/${id}`,
-    );
-    if (response.ok) {
-      const location = await response.json();
-      if (location) return location.data;
+
+  private async getRequired(path: string): Promise<any> {
+    const url = `${this.apiOrigin}${this.basePath}${path}`;
+    const response = await fetch(url);
+
+    if (!response.ok) {
+      const payload = await response.text();
+      const message = `Request failed with ${response.status} ${response.statusText}, ${payload}`;
+      throw new Error(message);
     }
-    return undefined;
+
+    return await response.json();
   }
+
+  private async getOptional(path: string): Promise<any | undefined> {
+    const url = `${this.apiOrigin}${this.basePath}${path}`;
+    const response = await fetch(url);
+
+    if (!response.ok) {
+      if (response.status === 404) {
+        return undefined;
+      }
+
+      const payload = await response.text();
+      const message = `Request failed with ${response.status} ${response.statusText}, ${payload}`;
+      throw new Error(message);
+    }
+
+    return await response.json();
+  }
+
+  async getLocationById(id: String): Promise<Location | undefined> {
+    return await this.getOptional(`/locations/${id}`);
+  }
+
   async getEntities(
     filter?: Record<string, string>,
   ): Promise<DescriptorEnvelope[]> {
-    let url = `${this.apiOrigin}${this.basePath}/entities`;
+    let path = `/entities`;
     if (filter) {
-      url += '?';
-      url += Object.entries(filter)
+      path += '?';
+      path += Object.entries(filter)
         .map(
           ([key, value]) =>
             `${encodeURIComponent(key)}=${encodeURIComponent(value)}`,
         )
         .join('&');
     }
-    const response = await fetch(url);
-    if (!response.ok) {
-      const payload = await response.text();
-      throw new Error(
-        `Request failed with ${response.status} ${response.statusText}, ${payload}`,
-      );
-    }
 
-    return await response.json();
+    return await this.getRequired(path);
   }
 
-  async getEntity({
-    name,
-    namespace,
-    kind,
-  }: {
-    name: string;
-    namespace?: string;
-    kind: string;
-  }): Promise<DescriptorEnvelope> {
-    const response = await fetch(
-      `${this.apiOrigin}${this.basePath}/entities/by-name/${kind}/${
-        namespace ?? 'default'
-      }/${name}`,
-    );
-    const entity = await response.json();
-    if (entity) return entity;
-    throw new Error(`'Entity not found: ${name}`);
+  async getEntityByName(
+    compoundName: EntityCompoundName,
```

Comment:
Still could've been destructed here, right?

Reply:
Sure but I don't prefer it. When call signatures become too complex and mix several different actions AND have hidden argument names, I feel it can hurt readability.

Reply:
(just clarifying that it's a personal taste and i don't mind anybody doing differently, it was just how i ended up typing things because it's how i tend to think about it)